lib/hls/hls_parser.js

```diff
@@ -2899,59 +2899,59 @@ shaka.hls.HlsParser = class {
       }
     }
 
-    const aesKeyInfoKey = `${drmTag.toString()}-${firstMediaSequenceNumber}`;
-    if (!this.aesKeyInfoMap_.has(aesKeyInfoKey)) {
-      // Default AES-128
-      const keyInfo = {
-        bitsKey: 128,
-        blockCipherMode: 'CBC',
-        iv,
-        firstMediaSequenceNumber,
-      };
-
-      const method = drmTag.getRequiredAttrValue('METHOD');
-      switch (method) {
-        case 'AES-256':
-          keyInfo.bitsKey = 256;
-          break;
-        case 'AES-256-CTR':
-          keyInfo.bitsKey = 256;
-          keyInfo.blockCipherMode = 'CTR';
-          break;
-      }
-
-      // Don't download the key object until the segment is parsed, to avoid a
-      // startup delay for long manifests with lots of keys.
-      keyInfo.fetchKey = async () => {
-        const keyUris = shaka.hls.Utils.constructSegmentUris(
-            getUris(), drmTag.getRequiredAttrValue('URI'), variables);
+    // Default AES-128
+    const keyInfo = {
+      bitsKey: 128,
+      blockCipherMode: 'CBC',
+      iv,
+      firstMediaSequenceNumber,
+    };
 
+    const method = drmTag.getRequiredAttrValue('METHOD');
+    switch (method) {
+      case 'AES-256':
+        keyInfo.bitsKey = 256;
+        break;
+      case 'AES-256-CTR':
+        keyInfo.bitsKey = 256;
+        keyInfo.blockCipherMode = 'CTR';
+        break;
+    }
+
+    // Don't download the key object until the segment is parsed, to avoid a
+    // startup delay for long manifests with lots of keys.
+    keyInfo.fetchKey = async () => {
+      const keyUris = shaka.hls.Utils.constructSegmentUris(
+          getUris(), drmTag.getRequiredAttrValue('URI'), variables);
+
+      const keyMapKey = keyUris.sort().join('');
+      if (!this.aesKeyMap_.has(keyMapKey)) {
         const requestType = shaka.net.NetworkingEngine.RequestType.KEY;
         const request = shaka.net.NetworkingEngine.makeRequest(
             keyUris, this.config_.retryParameters);
         const keyResponse =
             await this.makeNetworkRequest_(request, requestType);
+        this.aesKeyMap_.set(keyMapKey, keyResponse.data);
```

Comment:
Technically there could still be duplicate AES requests if a key is requested for a second time while the key is still downloading, right?
This could be structured as a map of promise objects instead, perhaps, that are set when the key is requested and resolve when the download is finished.

Reply:
In my tests that adds latency.. :(

Reply:
Done!